kernel/comps/keyboard/src/i8042_chip/keyboard.rs

```diff
@@ -0,0 +1,373 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! The i8042 keyboard driver.
+
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use ostd::{
+    arch::kernel::{MappedIrqLine, IRQ_CHIP},
+    trap::{IrqLine, TrapFrame},
+};
+use spin::Once;
+
+use super::controller::{I8042Controller, I8042ControllerError, I8042_CONTROLLER};
+use crate::{InputKey, KEYBOARD_CALLBACKS};
+
+/// IRQ line for i8042 keyboard.
+static IRQ_LINE: Once<MappedIrqLine> = Once::new();
+
+/// ISA interrupt number for i8042 keyboard.
+const ISA_INTR_NUM: u8 = 1;
+
+pub(super) fn init(controller: &mut I8042Controller) -> Result<(), I8042ControllerError> {
+    // Reset keyboard device by sending 0xFF (reset command, supported by all PS/2 devices) to port 1
+    // and waiting for a response.
+    controller.wait_and_send_data(0xFF)?;
+
+    // If the response is 0xFA, 0xAA (Note: the order in which devices send these two seems to be ambiguous)
+    // followed by the device PS/2 ID, then the device reset was successful.
+    //
+    // See <https://wiki.osdev.org/I8042_PS/2_Controller#Detecting_PS/2_Device_Types> for a list of IDs.
+    let mut iter = core::iter::from_fn(|| controller.wait_and_recv_data().ok());
+    match (iter.next(), iter.next()) {
+        (Some(0xFA), Some(0xAA)) | (Some(0xAA), Some(0xFA)) => (),
+        _ => return Err(I8042ControllerError::DeviceResetFailed),
+    }
+    match (iter.next(), iter.next()) {
+        (None, None) => (),                                   // Ancient AT keyboard
+        _ => return Err(I8042ControllerError::DeviceUnknown), // Will our implementation work with other keyboards?
+    }
+
+    let mut irq_line = IrqLine::alloc().unwrap();
+    irq_line.on_active(handle_keyboard_input);
+    let Ok(mappped_irq_line) = IRQ_CHIP
+        .get()
+        .unwrap()
+        .map_isa_pin_to(irq_line, ISA_INTR_NUM)
+    else {
+        return Err(I8042ControllerError::DeviceAllocIrqFailed);
```

Comment:
Since this error is named as `DeviceAllocIrqFailed`, maybe return errors if `IrqLine::alloc` fails as well:
```rust
let irq_line = IrqLine::alloc()
    .and_then(|irq_line| IRQ_CHIP.get().unwrap().map_isa_pin_to(irq_line, ISA_INTR_NUM))
    .map_err(|_| I8042ControllerError::DeviceAllocIrqFailed)?;
irq_line.on_active(handle_keyboard_input);
IRQ_LINE.call_once(|| mappped_irq_line);
```